modules/core/src/lib/deck.ts

```diff
@@ -1024,6 +1024,7 @@ export default class Deck {
 
     setGLParameters(device, this.props.parameters);
 
+    this.device?.canvasContext?.resize({useDevicePixels: this.props.useDevicePixels});
```

Comment:
Not sure if this is correct place, but we must do it in one place or another, when we're rendering to canvas we're not "managing" by ourselves.

Reply:
After some research, i think the core problem is that luma.gl `CanvasContext.cssToDeviceRatio()` returns invalid value if no-one called `resize({useDevicePixels :true})` or `setDevicePixelRatio`  and it's not called by animation loop for external contexts.

I think best fix would be to fix `cssToDeviceRatio` so it works even if user didn't call `resize`.

Reply:
The `useDevicePixels` is being passed down to luma in `_createAnimationLoop`, which should be enough. I also note that there we have a TODO regarding `autoResizeDrawingBuffer` which may be related to my other comment:

https://github.com/visgl/deck.gl/blob/master/modules/core/src/lib/deck.ts#L803-L804

Reply:
The CSS pixel API on the CanvasContext is a bit confusing, and I think it can be a source of subtle errors. 

To create that API, I basically just took a bunch of global CSS pixel helper functions (that had been hacked on over the years) in luma.gl v8 and made them into members of the `CanvasContext`. 

While they fit well there, I feel these functions as written and documented right now are not easy to understand and there are some pitfalls with how drawing buffers and canvases relate in WebGL that require special handling so one has to be careful when refactoring them.

The existing test cases were also too difficult to port as they relied on hacky mocking of WebGL context so we lost that safety net. Such subtle functions without test cases is a bad situation.

For the call in question, I wonder if `useDevicePixels` wouldn't be better as a prop on the CanvasContext rather than a parameter to `resize()`. It kind of looks confusing to not send in width and height but to have to remember to send in useDevicePixels.